File: FacialRecognition/collect.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
)

func main() {
	fetchResultData("FaceDetectionMetrics")
}

func fetchResultData(tableName string) {
	file, err := os.OpenFile(fmt.Sprintf("%s.csv", tableName), os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0666)
	if err != nil {
		log.Fatal("Cannot create data file", err)
	}
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()
	writer.Write([]string{
		"duration",
		"maxRss",
		"fsRead",
		"fsWrite",
		"vContextSwitches",
		"ivContextSwitches",
		"userDiff",
		"sysDiff",
		"rss",
		"heapTotal",
		"heapUsed",
		"external",
		"elMin",
		"elMax",
		"elMean",
		"elStd",
		"bytecodeMetadataSize",
		"heapPhysical",
		"heapAvailable",
		"heapLimit",
		"mallocMem",
		"netByRx",
		"netPkgRx",
		"netByTx",
		"netPkgTx",
	})
	sess := session.Must(session.NewSession())
	dynamoClient := dynamodb.New(sess, aws.NewConfig().WithRegion("eu-west-1"))
	dynamoClient.ScanPages(&dynamodb.ScanInput{
		TableName: &tableName,
	}, func(result *dynamodb.ScanOutput, lastPage bool) bool {
		for _, response := range result.Items {
			writer.Write([]string{
				*response["duration"].N,
				*response["maxRss"].N,
				*response["fsRead"].N,
				*response["fsWrite"].N,
				*response["vContextSwitches"].N,
				*response["ivContextSwitches"].N,
				*response["userDiff"].N,
				*response["sysDiff"].N,
				*response["rss"].N,
				*response["heapTotal"].N,
				*response["heapUsed"].N,
				*response["external"].N,
				*response["elMin"].N,
				*response["elMax"].N,
				*response["elMean"].N,
				*response["elStd"].N,
				*response["bytecodeMetadataSize"].N,
				*response["heapPhysical"].N,
				*response["heapAvailable"].N,
				*response["heapLimit"].N,
				*response["mallocMem"].N,
				*response["netByRx"].N,
				*response["netPkgRx"].N,
				*response["netByTx"].N,
				*response["netPkgTx"].N,
			})
		}
		writer.Flush()
		return !lastPage
	})
}

File: AirlineBooking/src/backend/payment/src/collect-payment/collect.js
const fetch = require('node-fetch');
const AWS = require("aws-sdk");
const { v4: uuidv4 } = require('uuid');

let coldStart = true;

let paymentEndpoint = process.env.PAYMENT_API_URL;

async function collectPayment(chargeId, stripeKey) {
  if (!paymentEndpoint) {
    throw new Error("Payment API URL is invalid -- Consider reviewing PAYMENT_API_URL env");
  }
  let paymentPayload = { "chargeId": chargeId, "stripeKey": stripeKey };
  try {
    const ret = await fetch.default(paymentEndpoint, {
      method: "POST",
      body: JSON.stringify(paymentPayload),
      headers: { 'Content-Type': 'application/json' }
    });
    if (ret.status < 200 || ret.status >= 300) {
      throw new Error("Error occured!");
    }
    const paymentResponse = await ret.json();
    return {
      "receiptUrl": paymentResponse.capturedCharge.receipt_url,
      "price": paymentResponse.capturedCharge.amount
    }
  } catch (err) {
    throw err;
  }
}

async function lambdaHandler(event, context) {
  if (coldStart) {
    coldStart = true;
    console.log("COLDSTART", context.awsRequestId);
  }
  const preAuthorizationToken = event.chargeId;
  const customerId = event.customerId;
  const stripeKey = event.stripeKey;

  if (!preAuthorizationToken) {
    throw new Error("Invalid Charge ID");
  }
  try {
    const ret = collectPayment(preAuthorizationToken, stripeKey);
    return ret
  } catch (err) {
    throw err;
  }
}


// monitoring function wrapping arbitrary payload code
async function handler(event, context, payload) {
  const child_process = require("child_process");
  const v8 = require("v8");
  const { performance, PerformanceObserver, monitorEventLoopDelay } = require("perf_hooks");
  const [beforeBytesRx, beforePkgsRx, beforeBytesTx, beforePkgsTx] =
    child_process.execSync("cat /proc/net/dev | grep vinternal_1| awk '{print $2,$3,$10,$11}'").toString().split(" ");
  const startUsage = process.cpuUsage();
  const beforeResourceUsage = process.resourceUsage();
  const wrapped = performance.timerify(payload);
  const h = monitorEventLoopDelay();
  h.enable();

  const durationStart = process.hrtime();

  const ret = await wrapped(event, context);
  h.disable();

  const durationDiff = process.hrtime(durationStart);
  const duration = (durationDiff[0] * 1e9 + durationDiff[1]) / 1e6

  // Process CPU Diff
  const cpuUsageDiff = process.cpuUsage(startUsage);
  // Process Resources
  const afterResourceUsage = process.resourceUsage();

  // Memory
  const heapCodeStats = v8.getHeapCodeStatistics();
  const heapStats = v8.getHeapStatistics();
  const heapInfo = process.memoryUsage();

  // Network
  const [afterBytesRx, afterPkgsRx, afterBytesTx, afterPkgsTx] =
    child_process.execSync("cat /proc/net/dev | grep vinternal_1| awk '{print $2,$3,$10,$11}'").toString().split(" ");

  const dynamodb = new AWS.DynamoDB({ region: 'us-west-2' });
  if (!event.warmup)
    await dynamodb.putItem({
      Item: {
        "id": {
          S: uuidv4()
        },
        "duration": {
          N: `${duration}`
        },
        "maxRss": {
          N: `${afterResourceUsage.maxRSS - beforeResourceUsage.maxRSS}`
        },
        "fsRead": {
          N: `${afterResourceUsage.fsRead - beforeResourceUsage.fsRead}`
        },
        "fsWrite": {
          N: `${afterResourceUsage.fsWrite - beforeResourceUsage.fsWrite}`
        },
        "vContextSwitches": {
          N: `${afterResourceUsage.voluntaryContextSwitches - beforeResourceUsage.voluntaryContextSwitches}`
        },
        "ivContextSwitches": {
          N: `${afterResourceUsage.involuntaryContextSwitches - beforeResourceUsage.involuntaryContextSwitches}`
        },
        "userDiff": {
          N: `${cpuUsageDiff.user}`
        },
        "sysDiff": {
          N: `${cpuUsageDiff.system}`
        },
        "rss": {
          N: `${heapInfo.rss}`
        },
        "heapTotal": {
          N: `${heapInfo.heapTotal}`
        },
        "heapUsed": {
          N: `${heapInfo.heapUsed}`
        },
        "external": {
          N: `${heapInfo.external}`
        },
        "elMin": {
          N: `${h.min}`
        },
        "elMax": {
          N: `${h.max}`
        },
        "elMean": {
          N: `${isNaN(h.mean) ? 0 : h.mean}`
        },
        "elStd": {
          N: `${isNaN(h.stddev) ? 0 : h.stddev}`
        },
        "bytecodeMetadataSize": {
          N: `${heapCodeStats.bytecode_and_metadata_size}`
        },
        "heapPhysical": {
          N: `${heapStats.total_physical_size}`
        },
        "heapAvailable": {
          N: `${heapStats.total_available_size}`
        },
        "heapLimit": {
          N: `${heapStats.heap_size_limit}`
        },
        "mallocMem": {
          N: `${heapStats.malloced_memory}`
        },
        "netByRx": {
          N: `${afterBytesRx - beforeBytesRx}`
        },
        "netPkgRx": {
          N: `${afterPkgsRx - beforePkgsRx}`
        },
        "netByTx": {
          N: `${afterBytesTx - beforeBytesTx}`
        },
        "netPkgTx": {
          N: `${afterPkgsTx - beforePkgsTx}`
        }
      },
      TableName: "long.ma.collect-payment-metrics"
    }).promise();

  return ret;
};

exports.handler = async (event, context) => {
  return await handler(event, context, lambdaHandler);
}
